### PR TITLE
Adds type hints and jaxtyping throughout PhysicsNeMo-Mesh

### PR DESCRIPTION
**Pull request by peterdsharpe**
# PhysicsNeMo Pull Request

## Description

Adds and overhauls type hints (including jaxtyping where possible, and casting for type-checker narrowing) and docstrings (aligning to proper Sphinx style, and fixing some non-NumPy-format docstrings) throughout PhysicsNeMo-Mesh.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [x] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/mesh/calculus/divergence.py [hunk 1]
```diff
@@ -19,7 +19,8 @@
 Implements divergence using both DEC and LSQ methods.
 
 The DEC divergence is the composition ⋆₀⁻¹ d* ⋆₁ ♭(X), where ♭ is the
```
Comment (laserkelvin) [inline physicsnemo/mesh/calculus/divergence.py]: I was going to point out that the weird unicode might have difficulty rendering correctly in sphinx, but not sure if it's just the GitHub UI but `♭(X)` is like...a flat X?

File: physicsnemo/mesh/calculus/README.md [hunk 2]
```diff
@@ -229,32 +241,36 @@ Otherwise: |⋆v| = area(T)/4
...
 
 **Property**: Perfect tiling: `Σ_vertices |⋆v| = |mesh|` (conservation holds
 exactly)
 
 **References**:
 
-- Meyer et al. (2003) Sections 3.2-3.4
-- Desbrun et al. (2005) lines 286-395
-- Hirani (2003) Def. 2.4.5
+- Meyer et al. (2003), *Discrete Differential-Geometry Operators for Triangulated
```
Comment (laserkelvin) [inline physicsnemo/mesh/calculus/README.md]: Not sure how much you care, but you could use footnote syntax (e.g. `[^1]: Meyer et al. ...`) to make the in-lines a bit more compact

File: physicsnemo/mesh/curvature/_utils.py [hunk 3]
```diff
@@ -42,16 +43,17 @@ def compute_full_angle_n_sphere(n_manifold_dims: int) -> float:
...
         - nD: 2π^(n/2) / Γ(n/2) for n ≥ 2
 
     Examples
     --------
-        >>> import math
-        >>> assert abs(compute_full_angle_n_sphere(1) - math.pi) < 1e-10  # π
-        >>> assert abs(compute_full_angle_n_sphere(2) - 2*math.pi) < 1e-5  # 2π
+    >>> import math
+    >>> assert abs(compute_full_angle_n_sphere(1) - math.pi) < 1e-10  # π
+    >>> assert abs(compute_full_angle_n_sphere(2) - 2*math.pi) < 1e-5  # 2π
```
Comment (laserkelvin) [inline physicsnemo/mesh/curvature/_utils.py]: Is the different in numerical precision meaningful?

File: physicsnemo/mesh/geometry/_cell_normals.py [hunk 4]
```diff
@@ -46,27 +46,31 @@ def compute_cell_normals(
...
+    Parameters
+    ----------
+    relative_vectors : torch.Tensor
+        Edge vectors of shape ``(n_cells, n_manifold_dims, n_spatial_dims)``.
+        Row ``i`` is the vector from vertex 0 to vertex ``i+1`` of each simplex.
+        Must satisfy ``n_manifold_dims == n_spatial_dims - 1``.
+
+    Returns
+    -------
+    torch.Tensor
```
Comment (laserkelvin) [inline physicsnemo/mesh/geometry/_cell_normals.py]: Change the return signature to match the hint

File: physicsnemo/mesh/geometry/dual_meshes.py [hunk 5]
```diff
@@ -233,65 +241,68 @@ def _compute_meyer_mixed_voronoi_areas(
...
 
-    Dimension-specific algorithms:
+        Dimension-specific algorithms:
 
-    **1D manifolds (edges)**:
-        Each vertex receives half the length of each incident edge.
-        Formula: V(v) = Σ_{edges ∋ v} |edge|/2
+        **1D manifolds (edges)**:
+            Each vertex receives half the length of each incident edge.
+            Formula: V(v) = Σ_{edges ∋ v} |edge|/2
```
Comment (laserkelvin) [inline physicsnemo/mesh/geometry/dual_meshes.py]: Might want to consider using reST math for this function as a whole to help the notation render correctly since this is public API

File: physicsnemo/mesh/geometry/dual_meshes.py [hunk 6]
```diff
@@ -333,7 +349,9 @@ def compute_dual_volumes_0(mesh: "Mesh") -> Float[torch.Tensor, " n_points"]:
 
     elif n_manifold_dims == 2:
         ### 2D: Mixed Voronoi area for triangles using Meyer et al. 2003 algorithm
-        # Reference: Section 3.3 (Equation 7) and Section 3.4 (Figure 4)
+        # Reference: Meyer et al. (2003), "Discrete Differential-Geometry
```
Comment (laserkelvin) [inline physicsnemo/mesh/geometry/dual_meshes.py]: I think some of these inline comments should go in the docstring instead

File: physicsnemo/mesh/geometry/dual_meshes.py [hunk 7]
```diff
@@ -433,36 +452,36 @@ def compute_circumcenters(
...
         use least-squares.
     """
-    n_simplices, n_vertices, n_spatial_dims = vertices.shape
-    n_manifold_dims = n_vertices - 1
+    n_cells, n_verts_per_cell, n_spatial_dims = vertices.shape
+    n_manifold_dims = n_verts_per_cell - 1
 
     ### Handle special cases
-    if n_vertices == 1:
+    if n_verts_per_cell == 1:
```
Comment (laserkelvin) [inline physicsnemo/mesh/geometry/dual_meshes.py]: Maybe `match/case` here instead of if/else

File: physicsnemo/mesh/geometry/dual_meshes.py [hunk 8]
```diff
@@ -471,14 +490,14 @@ def compute_circumcenters(
...
             # Solve A @ x = b
             c_minus_v0 = torch.linalg.solve(
-                A,  # (n_simplices, n_dims, n_dims)
-                b.unsqueeze(-1),  # (n_simplices, n_dims, 1)
-            ).squeeze(-1)  # (n_simplices, n_dims)
+                A,  # (n_cells, n_spatial_dims, n_spatial_dims)
+                b.unsqueeze(-1),  # (n_cells, n_spatial_dims, 1)
+            ).squeeze(-1)  # (n_cells, n_spatial_dims)
         except torch.linalg.LinAlgError:
             # Singular matrix - fall back to least squares
```
Comment (laserkelvin) [inline physicsnemo/mesh/geometry/dual_meshes.py]: Instead of `try/except` for this, could compute it this way without graph breaks:

```python
if n_manifold_dims == n_spatial_dims:
    solve_solution, info = torch.linalg.solve_ex(A, rhs, check_errors=False)
    solve_solution = solve_solution.squeeze(-1)

    lstsq_solution = torch.linalg.lstsq(A, b.unsqueeze(-1)).solution.squeeze(-1)

    mask = info.ne(0).unsqueeze(-1)
    c_minus_v0 = torch.where(mask, lstsq_solution, solve_solution)
else:
    c_minus_v0 = torch.linalg.lstsq(A, b.unsqueeze(-1)).solution.squeeze(-1)
```

File: physicsnemo/mesh/geometry/interpolation.py [hunk 9]
```diff
@@ -52,7 +54,8 @@ def compute_barycentric_gradients(
     of the barycentric interpolation function that is 1 at vertex v and 0 at
     all other vertices of the cell.
 
-    These gradients are needed for the PP-sharp operator (Hirani Eq. 5.8.1).
+    These gradients are needed for the PP-sharp operator (Hirani 2003,
```
Comment (laserkelvin) [inline physicsnemo/mesh/geometry/interpolation.py]: @peterdsharpe operator!

File: physicsnemo/mesh/geometry/support_volumes.py [hunk 10]
```diff
@@ -54,13 +57,16 @@ def compute_edge_support_volume_cell_fractions(
     For each edge and each cell containing it, computes the fraction of the edge's
     dual 1-cell (and support volume) that lies within that cell.
 
-    This is needed for the DPP-flat operator (Hirani Eq. 5.5.3, line 2398):
+    This is needed for the DPP-flat operator (Hirani 2003, *Discrete Exterior
+    Calculus*, Equation 5.5.3):
         ⟨X♭, edge⟩ = Σ_{cells ⊃ edge} (|⋆edge ∩ cell|/|⋆edge|) × X(cell) · edge⃗
```
Comment (laserkelvin) [inline physicsnemo/mesh/geometry/support_volumes.py]: Rewrite as reST math

File: physicsnemo/mesh/io/io_pyvista.py [hunk 11]
```diff
@@ -548,24 +552,28 @@ def _build_dual_graph_edges(
...
-        The lazily-imported pyvista module.
 
     Returns
     -------
     int
         0, 1, 2, or 3.
     """
+    import importlib
+
+    pv = importlib.import_module("pyvista")
```
Comment (laserkelvin) [inline physicsnemo/mesh/io/io_pyvista.py]: Is there a reason why the import can't go to the top of the module?

File: physicsnemo/mesh/utilities/_edge_lookup.py [hunk 12]
```diff
@@ -27,9 +27,9 @@
...
 def find_edges_in_reference(
     reference_edges: Int[torch.Tensor, "n_ref 2"],
-    query_edges: Int[torch.Tensor, "n_query 2"],
+    query_edges: Int[torch.Tensor, "n_queries 2"],
     index_bound: int | None = None,
-) -> tuple[Int[torch.Tensor, " n_query"], Bool[torch.Tensor, " n_query"]]:
+) -> tuple[Int[torch.Tensor, " n_queries"], Bool[torch.Tensor, " n_queries"]]:
     """Find indices of query edges within a reference edge set.
 
     Uses hash-based lookup with O(n log n) complexity for sorting
```
Comment (laserkelvin) [inline physicsnemo/mesh/utilities/_edge_lookup.py]: For consistency, do you want to make this reST math as well?

File: physicsnemo/mesh/domain_mesh.py [hunk 13]
```diff
@@ -768,7 +772,7 @@ def validate(
         check_manifoldness: bool = False,
         tolerance: float = 1e-10,
         raise_on_error: bool = False,
-    ) -> dict:
+    ) -> Mapping[str, Any]:
```
Comment (laserkelvin) [inline physicsnemo/mesh/domain_mesh.py]: I'm not a huge fan of `Any`, and wondering whether `Mapping` is the right hint considering your docstring, over something like `TypedDict` (since you have known keys)

File: physicsnemo/mesh/domain_mesh.py [hunk 14]
```diff
@@ -988,7 +995,7 @@ def draw(
         boundary_kwargs: dict[str, Any] | None = None,
         ax: Any = None,
         backend_options: dict[str, Any] | None = None,
-    ):
+    ) -> Any:
```
Comment (laserkelvin) [inline physicsnemo/mesh/domain_mesh.py]: Misaligned with the docstring, which is meant to be an `matplotlib` object?

**Review by laserkelvin — CHANGES_REQUESTED** [review]
Generally LGTM, but I've added some points pertaining to how docstrings render, and one or two cases where I think the annotations don't align

File: physicsnemo/mesh/calculus/divergence.py (hunk 1 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/calculus/divergence.py]: Haha, this is actually correct syntax using *musical notation* math: https://en.wikipedia.org/wiki/Musical_isomorphism

File: physicsnemo/mesh/geometry/_cell_normals.py (hunk 4 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/geometry/_cell_normals.py]: Whoops haha, good catch

File: physicsnemo/mesh/geometry/interpolation.py (hunk 9 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/geometry/interpolation.py]: lol

File: physicsnemo/mesh/calculus/divergence.py (hunk 1 above)
Comment (laserkelvin) [reply physicsnemo/mesh/calculus/divergence.py]: That's fair - but just wanted to make sure that if you want it rendered correctly (at least not have sphinx be angry at whoever is building docs) you'll want it in reST math notation if you can, and not the unicode

Comment (peterdsharpe) [reply physicsnemo/mesh/calculus/divergence.py]: Ah! Understood - will fix

File: physicsnemo/mesh/calculus/README.md (hunk 2 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/calculus/README.md]: Good idea, will do

File: physicsnemo/mesh/geometry/dual_meshes.py (hunk 5 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/geometry/dual_meshes.py]: Will fix

File: physicsnemo/mesh/geometry/dual_meshes.py (hunk 6 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/geometry/dual_meshes.py]: Will fix

File: physicsnemo/mesh/geometry/dual_meshes.py (hunk 7 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/geometry/dual_meshes.py]: Will fix

File: physicsnemo/mesh/geometry/dual_meshes.py (hunk 8 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/geometry/dual_meshes.py]: Hmmm now this is a cool idea! Thank you - will test this :)

File: physicsnemo/mesh/geometry/support_volumes.py (hunk 10 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/geometry/support_volumes.py]: Will fix

File: physicsnemo/mesh/io/io_pyvista.py (hunk 11 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/io/io_pyvista.py]: This was the recommended syntax as of the 2.0.0 refactor (to allow for importlinter) with optional dependencies. I think it still is - @coreyjadams to confirm?

Comment (coreyjadams) [reply physicsnemo/mesh/io/io_pyvista.py]: Actually I objected to your lower down import syntax @peterdsharpe and like imports at the top of the file!

But I am not a TOTAL control freak, and am willing to look the other way **sometimes**.

For optional imports, please use the `OptionalImport` syntax from physicsnemo/core.

Comment (peterdsharpe) [reply physicsnemo/mesh/io/io_pyvista.py]: Haha sounds good! Will change

File: physicsnemo/mesh/domain_mesh.py (hunk 14 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/domain_mesh.py]: Added narrowed return type

File: physicsnemo/mesh/utilities/_edge_lookup.py (hunk 12 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/utilities/_edge_lookup.py]: Converted for the docstring. Direct inline references (i.e., `n_queries` in code) are kept in code

File: physicsnemo/mesh/curvature/_utils.py (hunk 3 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/curvature/_utils.py]: No - made these symmetric at 1e-10.

File: physicsnemo/mesh/geometry/dual_meshes.py (hunk 8 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/geometry/dual_meshes.py]: Implemented.

File: physicsnemo/mesh/domain_mesh.py (hunk 13 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/domain_mesh.py]: Converting to TypedDict.

Comment (peterdsharpe) [reply physicsnemo/mesh/domain_mesh.py]: Actually, after implementing this, this gets a lot messier with TypedDict - I think I'd vote to keep it as-is. I will narrow it to `dict[str, Any]`, however.

**Review by laserkelvin — APPROVED** [review]
